rs-matter/src/transport/network/btp/context.rs

```diff
@@ -0,0 +1,378 @@
+    /// so that the peripheral can report to it subscribe, unsubscribe and write events.
+    pub(crate) fn on_event(&self, event: GattPeripheralEvent) {
+        let result = match event {
+            GattPeripheralEvent::NotifySubscribed(address) => self.on_subscribe(address),
+            GattPeripheralEvent::NotifyUnsubscribed(address) => self.on_unsubscribe(address),
+            GattPeripheralEvent::Write {
+                address,
+                data,
+                gatt_mtu,
+            } => self.on_write(address, data, gatt_mtu),
+        };
+
+        if let Err(e) = result {
+            error!("Unexpected error in GATT callback: {e:?}");
+        }
+    }
+
+    /// Handles a write event to characteristic `C1` from the GATT peripheral.
+    fn on_write(&self, address: BtAddr, data: &[u8], gatt_mtu: Option<u16>) -> Result<(), Error> {
+        trace!("Received {data:02x?} bytes from {address}");
+
+        self.sessions.lock(|sessions| {
+            let mut sessions = sessions.borrow_mut();
+
+            if Session::is_handshake(data)? {
+                if sessions.len() >= MAX_BTP_SESSIONS {
+                    warn!("Too many BTP sessions, dropping a handshake request from address {address}");
+                } else {
+                    // Unwrap is safe because we checked the length above
+                    sessions
+                        .push(Session::process_rx_handshake(address, data, gatt_mtu)?)
+                        .unwrap();
+                }
+
+                Ok(())
+            } else {
+                let Some(index) = sessions
+                    .iter_mut()
+                    .position(|session| session.address() == address)
+                else {
+                    warn!("Dropping data from address {address} because there is no session for it");
+                    return Ok(());
+                };
+
+                let session = &mut sessions[index];
+                let result = session.process_rx_data(data);
+
+                if result.is_err() {
+                    sessions.swap_remove(index);
+                    error!("Dropping session {address} because of an error: {result:?}");
+                }
+
+                self.available_notif.notify();
+                self.recv_notif.notify();
+                self.ack_notif.notify();
+                self.send_notif.notify();
+
+                result
+            }
+        })
+    }
+
+    /// Handles a subscribe event to characteristic `C2` from the GATT peripheral.
+    fn on_subscribe(&self, address: BtAddr) -> Result<(), Error> {
+        info!("Subscribe request from {address}");
+
+        self.sessions.lock(|sessions| {
+            let mut sessions = sessions.borrow_mut();
+            if let Some(session) = sessions
+                .iter_mut()
+                .find(|session| session.address() == address)
+            {
+                if !session.set_subscribed() {
+                    unreachable!();
```

Comment:
> please add comments on why some bits are unreachable for every use. Is this saying that the same BT address should never subscribe twice? 

Yes, the same BT address should never attempt a second subscription, _as long as it is still subscribed and its Gatt connection to us is still valid_. In terms of lower level operations, that would've meant that the peer had attempted a second cccd write on characteristic `C2`, which is invalid and should be captured by the `GattPeripheral` implementation.

But you are right that - in fact - we do not control this behavior. The other peer and the `GattPeripheral` controls it, so rather than issuing `unreachable!` here, I now return an `InvalidState` error.

> why do we assume this session is in State::new when subscribing?

Because that's how the BTP protocol is supposed to work, as described in the spec? (correct me if I'm wrong here):
- It all starts with a confirmed-`write` to C1 which SHOULD contain a Handshake Request packet. We create the session upon receive of the Handshake Request packet, and put the session in a `New` state
- _Without waiting for our answer to Handshake Req_, the peer should proceed by subscribing for notifications to our C2 characteristic (the cccd `write` thing) and we should therefore receive a `Subscribed` event. We should ONLY, EVER receive this for a session which is still in a `New` state, and only _once_. I.e. everything else is an invalid BTP protocol handshake
- Once we get the information that the other peer DID subscribe to characteristic C2, we can finally reply to their Handshake Request that did arrive earlier. We couldn't reply before that, because to reply, we need to have a valid subscription on the C2 characteristic from the same peer that did send us the Handshake Req, or else there is no way for us to "notify" the other peer (which is just a GATT-specific way to send to the peer a confirmed-write data). I.e. until we get a subscription to the peer, our "write" channel to the peer is not even open yet, so we cannot talk to it.

(Leaving as "Unresolved" in case you would like to correct my understanding of the spec.)